sklearn/preprocessing/data.py

```diff
@@ -1956,3 +1967,222 @@ def transform(self, X):
+
+    @staticmethod
+    def _valid_data_type(type_check):
+        """
+        Defines the data types that are compatible with CountFeaturizer
+        Currently, only Python lists and numpy arrays are accepted
+        """
+        return type(type_check) == np.ndarray or type(type_check) == list \
+            or type(type_check) == set
+
+    @staticmethod
+    def _check_well_formed(X):
+        """
+        Ensures that X is well formed, meaning that every single row of X must
+        have the same number of elements
+        [[1, 2], [3, 4]] is "well formed"
+        but [[1], [3, 5, 7]] is not "well formed"
+        Assumes that check_array(X) has already been called
+        """
+        # This is a temporary method but I am unsure whether there already
+        # exists something similar in the utils.validation.py
+        # There did not seem to be anything I could use to check well-formed
+        # in utils.validation.check_array()
+
+        num_columns = len(X[0])
+        for i in range(1, len(X)):
+            if len(X[i]) != num_columns:
+                raise ValueError("Malformed input array X")
+
+    def _extract_tuple(self, data_row):
+        """
+        Extracts the values of the data_row[inclusion] into an ordered tuple
+        for use as a dictionary key
+        """
+        if type(self.inclusion) == str and self.inclusion == 'all':
+            return tuple(data_row)
+        else:
+            num_features = len(data_row)
+            return tuple([data_row[i]
+                          for i in range(num_features)
+                          if (i in self.inclusion)])
+
+    def _final_num_features(self, cols_X):
+        """
+        Given the number of columns of X, uses the number of elements in
+        'inclusion' and 'removal_policy' to calclulate the number of
+        features in the output of transform()
+        """
+        if type(self.removal_policy) == str \
+                and self.removal_policy == 'inclusion':
+            if type(self.inclusion) == str and \
+                    self.inclusion == 'all':
+                return 1
+            else:
+                return cols_X - self.inclusion.size + 1
+        elif self.removal_policy is not None:
+            return cols_X - len(self.removal_policy) + 1
+        return cols_X + 1
+
+    def fit(self, X, y=None):
+        """
+        Sets the value of count_cache which holds the counts of each data point
+        """
+        X = check_array(X)
+        CountFeaturizer._check_well_formed(X)
+        cols_X = len(X[0])
+        if self._final_num_features(cols_X) <= 0:
+            # the removal_policy and inclusion were set such that the output
+            # array of transform() is expected to have 0 or a negative
+            # number of columns
+            raise ValueError("inclusion or removal_policy incompatible")
+        self.num_features = cols_X
+        self.count_cache = {}
```

Comment:
This is a model attribute and should end with `_`. Please read the contributors' guide.

perhaps use a `collections.Counter()`

Indeed, this could look something like `count_cache = Counter(tuple(x) for x in X.take(inclusion, axis=1).tolist())`